### ISSUE TITLE
VitTracker return all BB and scores

### ISSUE THREAD

**Issue by arielkantorovich** [opencv/opencv#24650]
### Describe the feature and motivation

It would be great if VitTracker would back all BBs and scores and not only the BB that corresponds to the highest score.
The main reasson is that VitTracker is very sensetivy to similar apperance so if you give the option (for example flag) that returns all BBs and scores we can add other post-processing algorithms using opencv that improve the results.

### Additional context

I tried to make this myself. I am more Python guy, I change the update function to:
`bool TrackerVitImpl::update(InputArray image_, std::vector<Rect>& boundingBoxes)
{
    image = image_.getMat().clone();
    Mat crop;
    crop_image(image, crop, rect_last, 4);
    Mat blob;
    preprocess(crop, blob, searchSize);
    net.setInput(blob, "search");
    std::vector<String> outputName = {"output1", "output2", "output3"};
    std::vector<Mat> outs;
    net.forward(outs, outputName);
    CV_Assert(outs.size() == 3);

    Mat conf_map = outs[0].reshape(0, {16, 16});
    Mat size_map = outs[1].reshape(0, {2, 16, 16});
    Mat offset_map = outs[2].reshape(0, {2, 16, 16});

    multiply(conf_map, (1.0 - hanningWindow), conf_map);

    // Store all bounding boxes and scores
    boundingBoxes.clear();
    for (int y = 0; y < conf_map.rows; ++y)
    {
        for (int x = 0; x < conf_map.cols; ++x)
        {
            float score = conf_map.at<float>(y, x);
            float cx = (x + offset_map.at<float>(0, y, x)) / 16;
            float cy = (y + offset_map.at<float>(1, y, x)) / 16;
            float w = size_map.at<float>(0, y, x);
            float h = size_map.at<float>(1, y, x);

            Rect boundingBox = returnfromcrop(cx - w / 2, cy - h / 2, w, h, rect_last);
            boundingBoxes.push_back(boundingBox);
        }
    }

    return true;
}`

but I get an error when I try to compile I would be happy to any help and of course, if you can add this as a flag to your code it will be awesome. thank you for a great job:)

**Comment by fengyuentau** [comment]
Ok, it is doable.

This feature can be implemented via overloading `bool update(InputArray image, CV_OUT std::vector<Rect>& boundingBox) = 0;`. In this case, we need to also implement this function for other trackers as well. What do you think? @asmorkalov @dkurt

**Comment by fengyuentau** [comment]
Actually it is more complicated as I try to implement such a feature. @arielkantorovich  What do you mean exactly by `other post-processing algorithms`? Could you list some examples on these algorithms?

---

Although we can always return all bboxes, we also need to update the box for tracking. Instead of overloading `bool update(InputArray image, CV_OUT std::vector<Rect>& boundingBox) = 0;`, we need to extract it into two functions, one is getting all boxes, the other one is update the tracking bbox.

**Comment by arielkantorovich** [comment]
@fengyuentau 
I add here some example of post-processing technique [https://arxiv.org/abs/2211.06663](url)
I going to implement something more simple that needs fewer computation resources but the main idea is from this paper.
I do not relly understand from your answer what is missing in my code I add the declaration  `update(InputArray image, CV_OUT std::vector<Rect>& boundingBox) = 0;`, I need to add another function?

**Comment by arielkantorovich** [comment]
@fengyuentau 
I will be happy if you can give me some instructions on what need to add to my code such I can get all BB.

**Comment by fengyuentau** [comment]
This feature looks more like a personal need to me. If you do need this, I think you should replace the post-processing algorithm with your own or submit a pull request for the replacement if you do prove that yours is better than the current one.

**Comment by arielkantorovich** [comment]
Ok, first for implement my post processing I need get all BB I attached my code above there is addition think that I need to implement ? I will be happy to any help

**Comment by fengyuentau** [comment]
Your code looks good to me roughly. I suggest you try your code first then report issues here if it does not return all bboxes.

**Comment by arielkantorovich** [comment]
I get this error when I try to build the new code:
Error:
`/mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/tracker_vit.cpp:54:10: error: initializer specified for non-virtual method ‘bool cv::TrackerV
itImpl::update(cv::InputArray, std::vector<cv::Rect_<int> >&)’
     bool update(InputArray image, CV_OUT std::vector<Rect>& boundingBoxes) = 0;
          ^
/mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/tracker_vit.cpp: In member function ‘bool cv::TrackerVitImpl::update(cv::InputArray, std::vec
tor<cv::Rect_<int> >&)’:
/mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/tracker_vit.cpp:195:19: warning: unused variable ‘score’ [-Wunused-variable]
             float score = conf_map.at<float>(y, x);
                   ^
In file included from /usr/include/x86_64-linux-gnu/c++/5/bits/c++allocator.h:33:0,
                 from /usr/include/c++/5/bits/allocator.h:46,
                 from /usr/include/c++/5/string:41,
                 from /usr/include/c++/5/stdexcept:39,
                 from /usr/include/c++/5/array:38,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/core/include/opencv2/core/cvdef.h:794,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/core/include/opencv2/core.hpp:52,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/include/opencv2/video/tracking.hpp:47,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/include/opencv2/video.hpp:56,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/../precomp.hpp:46,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/tracker_vit.cpp:7:
/usr/include/c++/5/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = cv::TrackerVitImpl; _Args = {const cv::Tracker
Vit::Params&}; _Tp = cv::TrackerVitImpl]’:
/usr/include/c++/5/bits/alloc_traits.h:530:4:   required from ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::all
ocator_type&, _Up*, _Args&& ...) [with _Up = cv::TrackerVitImpl; _Args = {const cv::TrackerVit::Params&}; _Tp = cv::TrackerVitImpl; std::allocator_traits<std::allocator<_CharT> >::allo
cator_type = std::allocator<cv::TrackerVitImpl>]’
/usr/include/c++/5/bits/shared_ptr_base.h:522:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {const cv:
:TrackerVit::Params&}; _Tp = cv::TrackerVitImpl; _Alloc = std::allocator<cv::TrackerVitImpl>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:617:4:   required from ‘std::__shared_count<_Lp>::__shared_count(std::_Sp_make_shared_tag, _Tp*, const _Alloc&, _Args&& ...) [with _Tp = cv::T
rackerVitImpl; _Alloc = std::allocator<cv::TrackerVitImpl>; _Args = {const cv::TrackerVit::Params&}; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr_base.h:1097:35:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::
allocator<cv::TrackerVitImpl>; _Args = {const cv::TrackerVit::Params&}; _Tp = cv::TrackerVitImpl; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2u]’
/usr/include/c++/5/bits/shared_ptr.h:319:64:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<cv::T
rackerVitImpl>; _Args = {const cv::TrackerVit::Params&}; _Tp = cv::TrackerVitImpl]’
/usr/include/c++/5/bits/shared_ptr.h:620:39:   required from ‘std::shared_ptr<_Tp1> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = cv::TrackerVitImpl; _Alloc = std::alloc
ator<cv::TrackerVitImpl>; _Args = {const cv::TrackerVit::Params&}]’
/usr/include/c++/5/bits/shared_ptr.h:635:39:   required from ‘std::shared_ptr<_Tp1> std::make_shared(_Args&& ...) [with _Tp = cv::TrackerVitImpl; _Args = {const cv::TrackerVit::Params&
}]’
/mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/core/include/opencv2/core/cvstd_wrapper.hpp:146:43:   required from ‘cv::Ptr<_Tp> cv::makePtr(const A1& ...) [wi
th _Tp = cv::TrackerVitImpl; A1 = {cv::TrackerVit::Params}]’
/mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/tracker_vit.cpp:216:46:   required from here
/usr/include/c++/5/ext/new_allocator.h:120:4: error: invalid new-expression of abstract class type ‘cv::TrackerVitImpl’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^
/mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/tracker_vit.cpp:40:7: note:   because the following virtual functions are pure within ‘cv::Tr
ackerVitImpl’:
 class TrackerVitImpl : public TrackerVit
       ^
In file included from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/include/opencv2/video.hpp:56:0,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/../precomp.hpp:46,
                 from /mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/src/tracking/tracker_vit.cpp:7:
/mnt/c/Users/arielka/PycharmProjects/Opencv_Basic_course/opencv/modules/video/include/opencv2/video/tracking.hpp:737:10: note:  virtual bool cv::Tracker::update(cv::InputArray, cv::Rec
t&)
     bool update(InputArray image, CV_OUT Rect& boundingBox) = 0;
          ^
modules/video/CMakeFiles/opencv_video.dir/build.make:647: recipe for target 'modules/video/CMakeFiles/opencv_video.dir/src/tracking/tracker_vit.cpp.o' failed
make[2]: *** [modules/video/CMakeFiles/opencv_video.dir/src/tracking/tracker_vit.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
`


my code:
`class TrackerVit : public Tracker
{
public:
    // ...

    virtual bool update(InputArray image, CV_OUT std::vector<Rect>& boundingBoxes) = 0;

    // ...
};





bool TrackerVitImpl::update(InputArray image_, CV_OUT std::vector<Rect>& boundingBoxes)
{
    image = image_.getMat().clone();
    Mat crop;
    crop_image(image, crop, rect_last, 4);
    Mat blob;
    preprocess(crop, blob, searchSize);
    net.setInput(blob, "search");
    std::vector<String> outputName = {"output1", "output2", "output3"};
    std::vector<Mat> outs;
    net.forward(outs, outputName);
    CV_Assert(outs.size() == 3);

    Mat conf_map = outs[0].reshape(0, {16, 16});
    Mat size_map = outs[1].reshape(0, {2, 16, 16});
    Mat offset_map = outs[2].reshape(0, {2, 16, 16});

    multiply(conf_map, (1.0 - hanningWindow), conf_map);

    // Store all bounding boxes and scores
    boundingBoxes.clear();
    for (int y = 0; y < conf_map.rows; ++y)
    {
        for (int x = 0; x < conf_map.cols; ++x)
        {
            float score = conf_map.at<float>(y, x);
            float cx = (x + offset_map.at<float>(0, y, x)) / 16;
            float cy = (y + offset_map.at<float>(1, y, x)) / 16;
            float w = size_map.at<float>(0, y, x);
            float h = size_map.at<float>(1, y, x);

            Rect boundingBox = returnfromcrop(cx - w / 2, cy - h / 2, w, h, rect_last);
            boundingBoxes.push_back(boundingBox);
        }
    }

    return true;
}`

**Comment by fengyuentau** [comment]
```
class TrackerVit : public Tracker
{
public:
    // ...

    virtual bool update(InputArray image, CV_OUT std::vector<Rect>& boundingBoxes) = 0;

    // ...
};
```

You need to add this to `class Tracker` instead. This is C++.

**Comment by arielkantorovich** [comment]
`class TrackerVitImpl : public TrackerVit
{
public:
    TrackerVitImpl(const TrackerVit::Params& parameters)
        : params(parameters)
    {
        net = dnn::readNet(params.net);
        CV_Assert(!net.empty());

        net.setPreferableBackend(params.backend);
        net.setPreferableTarget(params.target);
    }

    void init(InputArray image, const Rect& boundingBox) CV_OVERRIDE;
    bool update(InputArray image, CV_OUT std::vector<Rect>& boundingBoxes, CV_OUT std::vector<float>& scores) = 0;
    float getTrackingScore() CV_OVERRIDE;

    Rect rect_last;
    float tracking_score;

    TrackerVit::Params params;


protected:
    void preprocess(const Mat& src, Mat& dst, Size size);

    const Size searchSize{256, 256};
    const Size templateSize{128, 128};

    Mat hanningWindow;

    dnn::Net net;
    Mat image;
};`


that's is the right way?

**Comment by arielkantorovich** [comment]
@fengyuentau  sorry but I did not succeed in understanding from your answer what I need to do, there is an option that you give me more details?

**Comment by Daniel-Fundarek** [comment]
I tend to agree with @arielkantorovich, Vit tracker is missing options to retrieve all bboxes with confidence. Vit tracker doesn't get lost often, rather it starts to track whole screen or other object. However if i could access all bboxes and scores i could check for size and position similarity, use IOU or other association metrics. Thus reducing false target tracking.

**Comment by fengyuentau** [comment]
>  However if i could access all bboxes and scores i could check for size and position similarity, use IOU or other association metrics

Then you can look into the code the write your own tracking method.

**Comment by asmorkalov** [comment]
@fengyuentau could you take it in mean time?

**Comment by fengyuentau** [comment]
> @fengyuentau could you take it in mean time?

No. This feature request does not make sense to me. Trackers always pick the best bounding box without human intervention for the next prediction. It does not make sense to return all bounding boxes as we do not pause at every frame for a person to make choices on the boxes.

**Comment by TanNgocDo** [comment]
@arielkantorovich , have you implemented your algorithm successfully ?
